### PR TITLE
make the test runner fail if focused tests exist

### LINKED ISSUE #28: "focused" feature should be ignored in CI
Do we already have a cmd line switch for that so that FAKE runner can set it? /cc @vilinski

### PR DESCRIPTION
**Pull request by forki**
Fot the CI server we want to have an option to make the build fail if focused tests exist.

closes #28

### CODE REVIEW

**Comment by forki** [comment]
@haf what do you think?

**Comment by forki** [comment]
the broken CI doesn't look related. @Krzysztof-Cieslak what am I missing?

**Comment by Krzysztof-Cieslak** [comment]
Hmmm... not sure - https://travis-ci.org/haf/expecto/jobs/188576307#L1428-L1429

**Comment by forki** [comment]
ok that seems related. din't see that one

**Comment by Krzysztof-Cieslak** [comment]
Benchmark.Net errors are shown also on current master, and it's green

File: Expecto/Expecto.fs [hunk 1]
```diff
@@ -765,6 +765,9 @@ module Tests =
       /// true, because your code should not mutate global
       /// state by default.
       parallel : bool
+      /// Whether to make the test runner fail if focused tests exist.
+      /// This can be used from CI servers to ensure no focused tests are commited and therefor all tests are run.
+      fail_on_focused_tests : bool
```
Comment (haf) [inline Expecto/Expecto.fs]: Needs to be `camelCase`

Comment (forki) [inline Expecto/Expecto.fs]: fixed

File: Expecto.Tests/Main.fs [hunk 2]
```diff
@@ -1,7 +1,20 @@
...
 let main args =
+  let localList = 
+    testList "local" [
+      ftestList "focused" [
+        testCase "test" (fun () -> ())
+      ]
+    ]
+
+  // check if we can fail on focused tests
+  if runTests { defaultConfig with failOnFocusedTests = true } localList <> 1 then
```
Comment (haf) [inline Expecto.Tests/Main.fs]: Why can't we have this as a normal test?

Comment (forki) [inline Expecto.Tests/Main.fs]: Mhm I guess now we could have that. Will send ammendment or if you decide to merge then I will send another PR.

Comment (haf) [inline Expecto.Tests/Main.fs]: Amend, please.

**Comment by forki** [comment]
ok separated the tests

File: Expecto.Tests/FocusedTests.fs [hunk 3]
```diff
@@ -74,5 +74,31 @@ let all =
...
+  let localList = 
+    testList "local" [
+      testList "focused" [
+        testCase "test" (fun () -> ())
+      ]
+    ]
+
+  // check if we pass if no focused tests exist
+  if runTests { defaultConfig with failOnFocusedTests = true } localList <> 0 then
+    failwith "focused test check didn't fail"
```
Comment (haf) [inline Expecto.Tests/FocusedTests.fs]: Why not make it a normal test? This is code that executes on static module initialisation if I read the core correctly.

**Comment by haf** [comment]
Thanks @forki, it's looking better. Just the one niggling issue I commented about left.

File: Expecto.Tests/FocusedTests.fs (hunk 3 above)
Comment (forki) [inline Expecto.Tests/FocusedTests.fs]: ok pushed new version

**Comment by haf** [comment]
Also, now `runTests` is up from doing two things to doing three things. Can we separate those concerns a little?

**Comment by haf** [comment]
There's a bit of hanging whitespace, too. Can we remove that, please?
screen shot 2017-01-04 at 09 29 40

**Comment by forki** [comment]
ok I remove the trailing whitespace and added corresponding FSharpLint settings so that VS users see the issue.

**Comment by forki** [comment]
and extracted function so that runTests is small again.

File: Expecto/Expecto.fs [hunk 4]
```diff
@@ -765,6 +765,9 @@ module Tests =
       /// true, because your code should not mutate global
       /// state by default.
       parallel : bool
+      /// Whether to make the test runner fail if focused tests exist.
+      /// This can be used from CI servers to ensure no focused tests are commited and therefor all tests are run.
```
Comment (haf) [inline Expecto/Expecto.fs]: Max 80 char lines please

File: Expecto/Expecto.fs [hunk 5]
```diff
@@ -861,14 +865,35 @@ module Tests =
...
+    let count =
+      tests
+      |> List.filter (fun t ->
+          match t.state with
+          | Focused -> true
+          | _ -> false)
+      |> List.length
+
+    if count > 0 then
+      config.printer.info "It was requested that no focused tests exist, but yet there are %d focused tests found."
```
Comment (haf) [inline Expecto/Expecto.fs]: Where does `%d` come from?

Comment (haf) [inline Expecto/Expecto.fs]: Can we move logging out from the pure function?

File: Expecto/Expecto.fs [hunk 6]
```diff
@@ -861,14 +865,35 @@ module Tests =
     |> Test.toTestCodeList
     |> Seq.iter (fun t -> printfn "%s" t.name)
 
-  /// Runs tests with supplied options. Returns 0 if all tests passed, =
-  /// otherwise 1
-  let runTests config tests =
+  /// When the failOnFocusedTests switch is activated this function checks that no
+  /// focused tests exist. Returns true if the check passes, otherwise false.
+  let passesFocusTestCheck config tests =
+    if not config.failOnFocusedTests then true else
```
Comment (haf) [inline Expecto/Expecto.fs]: Can we move this check out to the caller and make it a disjunction?

File: Expecto/Expecto.fs (hunk 5 above)
Comment (forki) [inline Expecto/Expecto.fs]: fixed

File: Expecto/Expecto.fs [hunk 7]
```diff
@@ -879,7 +879,8 @@ module Tests =
       |> List.length
 
     if count > 0 then
-      config.printer.info "It was requested that no focused tests exist, but yet there are %d focused tests found."
+      sprintf "It was requested that no focused tests exist, but yet there are %d focused tests found." count
```
Comment (haf) [inline Expecto/Expecto.fs]: Could we possibly do it with the templating from Logary? `...but there are {count} focused tests found" >> setField "count" count)`?

File: Settings.FSharpLint [hunk 8]
```diff
@@ -1,10 +1,13 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <FSharpLintSettings xmlns="https://github.com/fsprojects/FSharpLint/blob/master/ConfigurationSchema.xsd">
-  <IgnoreFiles Update="Overwrite"><![CDATA[assemblyinfo.*
-]]></IgnoreFiles>
+  <IgnoreFiles Update="Overwrite"><![CDATA[assemblyinfo.*]]></IgnoreFiles>
```
Comment (haf) [inline Settings.FSharpLint]: Will lower-case assemblyinfo work on Linux?

Comment (forki) [inline Settings.FSharpLint]: fixed

File: Expecto/Expecto.fs (hunk 7 above)
Comment (forki) [inline Expecto/Expecto.fs]: done. but not sure if that's correct way. please review ca05aa6

**Comment by haf** [comment]
Thanks

**Pull request merged**